tesseract/consensus/polygon/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+use anyhow::Result;
+use codec::{Decode, Encode};
+use ismp::{
+	consensus::ConsensusStateId,
+	host::StateMachine,
+	messaging::{CreateConsensusState, Message},
+};
+use ismp_polygon::{ConsensusState, PolygonConsensusUpdate, POLYGON_CONSENSUS_CLIENT_ID};
+use serde::{Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use tendermint_primitives;
+use tendermint_prover::{Client, HeimdallClient};
+use tendermint_verifier::hashing::SpIoSha256;
+use tesseract_evm::{EvmClient, EvmConfig};
+use tesseract_primitives::{IsmpHost, IsmpProvider};
+
+mod notification;
+
+/// Host configuration for Polygon POS relayer
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct HostConfig {
+	/// Frequency (in seconds) to check for new updates
+	pub consensus_update_frequency: Option<u64>,
+	/// Heimdall RPC URL
+	pub heimdall_rpc_url: String,
+	/// Heimdall REST URL
+	pub heimdall_rest_url: String,
+	/// Execution RPC URL
+	pub execution_rpc_url: String,
+}
+
+/// Top-level config for Polygon POS relayer
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PolygonPosConfig {
+	pub host: HostConfig,
+	#[serde(flatten)]
+	pub evm_config: EvmConfig,
+}
+
+impl PolygonPosConfig {
+	pub async fn into_client(self) -> anyhow::Result<Arc<dyn IsmpHost>> {
+		Ok(Arc::new(PolygonPosHost::new(&self.host, &self.evm_config).await?))
+	}
+
+	pub fn state_machine(&self) -> StateMachine {
+		self.evm_config.state_machine
+	}
+}
+
+/// The relayer host for Polygon POS
+#[derive(Clone)]
+pub struct PolygonPosHost {
+	pub consensus_state_id: ConsensusStateId,
+	pub state_machine: StateMachine,
+	pub host: HostConfig,
+	pub evm: EvmConfig,
+	pub provider: Arc<dyn IsmpProvider>,
+	pub prover: HeimdallClient,
+}
+
+impl PolygonPosHost {
+	/// Create a new PolygonPosHost
+	pub async fn new(host: &HostConfig, evm: &EvmConfig) -> Result<Self, anyhow::Error> {
+		let ismp_provider = EvmClient::new(evm.clone()).await?;
+		Ok(Self {
+			consensus_state_id: {
+				let mut consensus_state_id: ConsensusStateId = Default::default();
+				consensus_state_id.copy_from_slice(evm.consensus_state_id.as_bytes());
+				consensus_state_id
+			},
+			state_machine: evm.state_machine,
+			host: host.clone(),
+			evm: evm.clone(),
+			provider: Arc::new(ismp_provider),
+			prover: HeimdallClient::new(
+				&host.heimdall_rpc_url,
+				&host.heimdall_rest_url,
+				&host.execution_rpc_url,
+			)?,
+		})
+	}
+
+	/// Fetch the current consensus state (for initial state creation)
+	pub async fn get_consensus_state(&self) -> Result<ConsensusState, anyhow::Error> {
+		let trusted_height = self.provider.query_finalized_height().await?.into();
```

Comment:
You are confusing heimdall queries with the evm queries.
To construct the initial consensus state, construct a heimdall trusted state by pulling the required data from heimdall, but for `consensus.state.last_finalized_block` and `consensus_state.last_finalized_hash` these are properties of the polygon evm chain, fetchthe latest milestone and fill inthiose values with milestone endblock and milestone hash

Reply:
Okay got it. What should be the trusted_height for heimdall?

Reply:
<img width="825" height="119" alt="image" src="https://github.com/user-attachments/assets/52aba809-390c-42e2-bffe-b9dedc871d86" />

Reply:
No, to get the initial tendermint trusted state state, fetch values from the handball rpc.

For the finalized height in the consensus state, use the end block in the latest milestone

Reply:
What should be the initial trusted height for tendermint? Latest height - 50?